community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java

```diff
@@ -131,7 +131,7 @@ TransactionType upgradeToSchemaTransaction() throws InvalidTransactionTypeKernel
     private TransactionHooks.TransactionHooksState hooksState;
     private KernelStatement currentStatement;
```

Comment:
These two have different concerns, enforce different things and throw different exceptions on failure. We did make the mistake in earlier versions of this PR of bringing these two closer together, but that was confusing the situation, so we've corrected that by separating them again. The TransactionType is now identical to what it was before, which is an internal transaction state tracking whether or not a transaction contains write or schema commands (and enforcing that it cannot contain both). This is entirely orthogonal to the authorisation concern of what a transaction is allowed to do. To keep this separation clear, we moved the AccessMode checks to the KernelStatement. The KernelTransactionImplementation is merely a carrier of the AccessMode. Enforcement of what is allowed is done in KernelStatement.